### PR TITLE
Whisper model support & `/v1/audio/transcriptions` endpoint & benchmark

### PR DESCRIPTION
**Pull request by JustinTong0323**
## Motivation
- Follow up #8064. Thanks @MahmoudAshraf97 for the brilliant work.
- This PR follows up the original PR and makes a workaround on the encoder-decoder code path.
- Adds the implementation of the `/v1/audio/transcriptions` endpoint (currently for whisper only).
- Also adds a benchmark script for the ASR task.

## Modifications
1. Support whisper model
2. Add a ASR benchmark

Further TODOs:
- [ ] Adapt to sglang's native encoder-decoder code path
- [ ] fix precision issue when using `--attention-backend flashinfer`

## Accuracy Tests
Launch Server:
```sh
sglang serve --model-path openai/whisper-large-v3
```
Run benchmark script:
```sh
cd sglang/benchmark/asr
python bench_sglang.py --api-type transcription --concurrency 128 --show-predictions
```
Output:
```sh
Loading dataset: D4nt3/esb-datasets-earnings22-validation-tiny-filtered...
Using API type: transcription
Repo card metadata block was not found. Setting CardData to empty.
WARNING:huggingface_hub.repocard:Repo card metadata block was not found. Setting CardData to empty.
Performing warmup...
Processing 511 samples...
------------------------------
Results for openai/whisper-large-v3:
Total Requests: 511
WER: 12.7690
Average Latency: 1.3602s
Median Latency: 1.2090s
95th Latency: 2.9986s
Throughput: 19.02 req/s
Token Throughput: 354.19 tok/s
Total Test Time: 26.8726s
------------------------------

==================== Sample Predictions ====================
Sample 1:
  REF: on the use of taxonomy i you know i think it is it is early days for us to to make any clear indications to the market about the proportion that would fall under that requirement
  PRED: on the eu taxonomy i think it is early days for us to make any clear indications to the market about the proportion that would fall under that requirement
----------------------------------------
Sample 2:
  REF: so within fiscal year 2021 say 120 a 100 depending on what the micro will do and next year it is not necessarily payable in q one is we will look at what the cash flows for 2022 look like
  PRED: so within fiscal year 2021 say $120000 $100000 depending on what the macro will do and next year it is not necessarily payable in q one is we will look at what the cash flows for 2022 look like
----------------------------------------
Sample 3:
  REF: we talked about 4.7 gigawatts
  PRED: we talked about 4.7 gigawatts
----------------------------------------
Sample 4:
  REF: and you know depending on that working capital build we will we will see what that yields
  PRED: and depending on that working capital build we will see what that yields what
----------------------------------------
Sample 5:
  REF: so on on sinopec what we have agreed with sinopec way back then is that free cash flows after paying all capexs are distributed out 30 70%
  PRED: so on sinopec what we have agreed with sinopec way back then is that free cash flows after paying all capexes are distributed out 30% 70%
----------------------------------------
============================================================
```

## Benchmarking and Profiling
As above.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/whisper.py [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+                if encoder_list:
+                    if len(encoder_list) == 1:
+                        encoder_outputs = encoder_list[0]
+                    else:
+                        # Concatenate encoder outputs for batched decode
+                        # This aligns with how decoder hidden states are concatenated
+                        encoder_outputs = torch.cat(encoder_list, dim=0)
+        else:
+            # Prefill (extend) phase: process each request's audio separately
+            encoder_list = []
```
Comment (mickqian) [inline python/sglang/srt/models/whisper.py:495]: is it possible that we use `general_mm_embed_routine`, since we'll be missing a lot of features if not

File: python/sglang/srt/models/whisper.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+            attn_output = attn_output.reshape(q_len, num_heads * head_dim)
+        else:
+            qkv, _ = self.qkv_proj(hidden_states)
+            q, k, v = qkv.chunk(chunks=3, dim=-1)
+            # Apply scaling to Q for self-attention
+            q = q * self.scaling
+
+            if self.is_encoder:
+                # For encoder: use direct attention (no KV cache, no RadixAttention)
+                # Encoder hidden_states may have batch dim: [batch, seq_len, embed_dim]
```
Comment (MahmoudAshraf97) [inline python/sglang/srt/models/whisper.py]: A valid usecase for whisper is batching audio inputs with variable lengths, if we can use an implementation that can handle padding it'd be great, although not a priority

File: python/sglang/srt/models/whisper.py (hunk 1 above)
Comment (JustinTong0323) [reply python/sglang/srt/models/whisper.py:495]: No(or need **much** more effort) as it's cross-attention model

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -1663,6 +1663,13 @@ def _handle_attention_backend_compatibility(self):
                 self.speculative_algorithm is None
             ), "Speculative decoding is currently not supported with Flex Attention backend"
 
+        # Encoder-decoder models (e.g., Whisper)
+        if model_config.is_encoder_decoder:
+            logger.warning(
+                "Cuda graph is disabled for encoder-decoder models (e.g., Whisper)"
+            )
+            self.disable_cuda_graph = True
```
Comment (yuan-luo) [inline python/sglang/srt/server_args.py:1789]: Just curious why disable cuda graph for encoder-decoder models?

Comment (JustinTong0323) [reply python/sglang/srt/server_args.py:1789]: some compatible issue with the current code path

**Review by yhyang201 — APPROVED** [review]

File: python/sglang/srt/models/whisper.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """Input shape: Batch x Time x Channel"""
+
+        if self.is_cross_attention:
+            q, _ = self.q_proj(hidden_states)
+            if cross_hidden_states is not None:
+                kv, _ = self.kv_proj(cross_hidden_states)
+                k, v = kv.split([self.kv_size, self.kv_size], dim=-1)
+            else:
+                k = torch.zeros_like(q)
```
Comment (vshekhawat-hlab) [inline python/sglang/srt/models/whisper.py:102]: Why we are having zero k and v. We will hit with junk output with this case, as k, v are not from encoder cache output. 

With flashinfer backend, are you getting any useful transcript output?
